merkle/compact/tree.go

```diff
@@ -191,8 +192,8 @@ func (t *Tree) AddLeaf(data []byte, visit VisitFn) ([]byte, error) {
 func (t *Tree) AddLeafHash(leafHash []byte, visit VisitFn) error {
 	defer func() {
 		t.size++
-		// TODO(pavelkalinnikov): Handle recalculateRoot errors.
-		t.recalculateRoot(visit)
+		// TODO(pavelkalinnikov): Handle calculateRoot errors.
```

Comment:
Since this is being recalculated in `CurrentRoot` do you still need this deferred func here? 
I guess yes because it's how intermediate nodes get written - but then is it necessary to recalculate it in `CurrentRoot`?

Is there a bigger picture that this PR is working toward?

Reply:
Correct, I keep it here only because all the code currently relies on the fact that `AddLeafHash` also reports **ephemeral** nodes via `visit` function. 

The bigger picture (see also PR description):
- `AddLeafHash` (I will probably rename it because of changed semantics) reports only non-ephemeral nodes, and thus calling `calculateRoot` is no longer necessary (key result: better performance).
- `calculateRoot` is exported and replaces `CurrentRoot`. It allows clients to compute root hash and get ephemeral nodes at the same time, as most of them need both operations anyway: e.g. the sequencer needs both after completing a series of `AddLeafHash` calls.
- Result: the API of `compact.Tree` is so close to `compact.Range` that replacing one with the other is a short PR.

Reply:
:+1: